cpp/src/parquet/geometry_util_internal.h

```diff
@@ -0,0 +1,395 @@
+      case 2:
+        return dimensions::XYM;
+      case 3:
+        return dimensions::XYZM;
+      default:
+        return ::arrow::Status::SerializationError("Invalid wkb_geometry_type: ",
+                                                   wkb_geometry_type);
+    }
+  }
+
+  static uint32_t size(dimensions dims) {
+    switch (dims) {
+      case dimensions::XY:
+        return 2;
+      case dimensions::XYZ:
+      case dimensions::XYM:
+        return 3;
+      case dimensions::XYZM:
+        return 4;
+      default:
+        return 0;
+    }
+  }
+
+  static std::string ToString(dimensions dims) {
+    switch (dims) {
+      case dimensions::XY:
+        return "XY";
+      case dimensions::XYZ:
+        return "XYZ";
+      case dimensions::XYM:
+        return "XYM";
+      case dimensions::XYZM:
+        return "XYZM";
+      default:
+        return "";
+    }
+  }
+};
+
+struct GeometryType {
+  enum class geometry_type {
+    POINT = 1,
+    LINESTRING = 2,
+    POLYGON = 3,
+    MULTIPOINT = 4,
+    MULTILINESTRING = 5,
+    MULTIPOLYGON = 6,
+    GEOMETRYCOLLECTION = 7
+  };
+
+  static ::arrow::Result<geometry_type> FromWKB(uint32_t wkb_geometry_type) {
+    switch (wkb_geometry_type % 1000) {
+      case 1:
+        return geometry_type::POINT;
+      case 2:
+        return geometry_type::LINESTRING;
+      case 3:
+        return geometry_type::POLYGON;
+      case 4:
+        return geometry_type::MULTIPOINT;
+      case 5:
+        return geometry_type::MULTILINESTRING;
+      case 6:
+        return geometry_type::MULTIPOLYGON;
+      case 7:
+        return geometry_type::GEOMETRYCOLLECTION;
+      default:
+        return ::arrow::Status::SerializationError("Invalid wkb_geometry_type: ",
+                                                   wkb_geometry_type);
+    }
+  }
+
+  static std::string ToString(geometry_type geometry_type) {
+    switch (geometry_type) {
+      case geometry_type::POINT:
+        return "POINT";
+      case geometry_type::LINESTRING:
+        return "LINESTRING";
+      case geometry_type::POLYGON:
+        return "POLYGON";
+      case geometry_type::MULTIPOINT:
+        return "MULTIPOINT";
+      case geometry_type::MULTILINESTRING:
+        return "MULTILINESTRING";
+      case geometry_type::MULTIPOLYGON:
+        return "MULTIPOLYGON";
+      case geometry_type::GEOMETRYCOLLECTION:
+        return "GEOMETRYCOLLECTION";
+      default:
+        return "";
+    }
+  }
+};
+
+class WKBBuffer {
+ public:
+  WKBBuffer() : data_(NULLPTR), size_(0) {}
+  WKBBuffer(const uint8_t* data, int64_t size) : data_(data), size_(size) {}
+
+  void Init(const uint8_t* data, int64_t size) {
+    data_ = data;
+    size_ = size;
+  }
+
+  ::arrow::Result<uint8_t> ReadUInt8() {
+    if (size_ < 1) {
+      throw ParquetException("Can't read 1 byte from empty WKBBuffer");
+    }
+
+    size_ -= 1;
+    return *data_++;
+  }
+
+  ::arrow::Result<uint32_t> ReadUInt32(bool swap) {
```

Comment:
is there a reason all class implementations are in the header? (holdover from templating)?

Reply:
Not an expert either, but I think as long as long as the .h/.cc file are well isolated I hope other implementations won't feel the need to reinvent the wheel (I guess using Status might be the primary detractor).

In terms of inlining, the rule of thumb is generally less then 10 lines of code in any particular function.